chatsky/script/core/context.py

```diff
@@ -110,155 +113,129 @@ class Context(BaseModel):
             self.misc.clear()
-        if "labels" in field_names:
-            for index in list(self.labels)[:-hold_last_n_indices]:
-                del self.labels[index]
         if "framework_data" in field_names:
             self.framework_data = FrameworkData()
 
+    async def delete(self) -> None:
+        if self._storage is not None:
+            await self._storage.delete_main_info(self.primary_id)
+        else:
+            raise RuntimeError(f"{type(self).__name__} is not attached to any context storage!")
+
+    @property
+    def labels(self) -> Dict[int, NodeLabel2Type]:
+        return {id: turn.label for id, turn in self.turns.items()}
+    
+    @property
+    def requests(self) -> Dict[int, Message]:
+        return {id: turn.request for id, turn in self.turns.items()}
+    
+    @property
+    def responses(self) -> Dict[int, Message]:
+        return {id: turn.response for id, turn in self.turns.items()}
+
+    def add_turn(self, turn: Turn):
+        last_index = get_last_index(self.turns)
+        self.turns[last_index + 1] = turn
+
+    def add_turn_items(self, label: Optional[NodeLabel2Type] = None, request: Optional[Message] = None, response: Optional[Message] = None):
+        request = Message() if request is None else request
+        response = Message() if response is None else response
+        self.add_turn(Turn(label=label, request=request, response=response))
+
+    @property
+    def last_turn(self) -> Optional[Turn]:
+        last_index = get_last_index(self.turns)
+        return self.turns.get(last_index)
+
+    @last_turn.setter
+    def last_turn(self, turn: Optional[Turn]):
+        last_index = get_last_index(self.turns)
+        self.turns[last_index] = Turn() if turn is None else turn
+
     @property
     def last_label(self) -> Optional[NodeLabel2Type]:
-        """
-        Return the last :py:data:`~.NodeLabel2Type` of
-        the :py:class:`~.Context`.
-        Return `None` if `labels` is empty.
+        return self.last_turn.label if self.last_turn is not None else None
 
-        Since `start_label` is not added to the `labels` field,
-        empty `labels` usually indicates that the current node is the `start_node`.
-        """
-        last_index = get_last_index(self.labels)
-        return self.labels.get(last_index)
+    @last_label.setter
+    def last_label(self, label: Optional[NodeLabel2Type]):
+        last_turn = self.last_turn
+        if last_turn is not None:
+            self.last_turn.label =  label
+        else:
+            raise ValueError("The turn history is empty!")
 
     @property
     def last_response(self) -> Optional[Message]:
-        """
-        Return the last `response` of the current :py:class:`~.Context`.
-        Return `None` if `responses` is empty.
-        """
-        last_index = get_last_index(self.responses)
-        return self.responses.get(last_index)
+        return self.last_turn.response if self.last_turn is not None else None
```

Comment:
This used to get response from previous turn if it wasn't generated yet.
I think we should update `last_response` and `last_label` to get data from previous turn if current turn doesn't have it.